### PR TITLE
Add support for Sagemaker Model Parallel >= 1.10 new checkpoint API

### PR DESCRIPTION
**Pull request by viclzhu**
# What does this PR do?
This PR adds support for Sagemaker Model Parallel >= 1.10's new checkpoint API as well as keeping SMP < 1.10 functionality.

* Support loading checkpoints saved with SMP < 1.10 in SMP < 1.10 and SMP >= 1.10

* Support loading checkpoints saved with SMP >= 1.10 in SMP >= 1.10

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -87,6 +90,12 @@
 
 _init_weights = True
 
+if is_sagemaker_mp_enabled():
+    import smdistributed.modelparallel.torch as smp
+    from smdistributed.modelparallel import __version__ as SMP_VERSION
+
+    SMP_POST_1_10 = version.parse(SMP_VERSION) >= version.parse("1.10")
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:97]:
```suggestion
    IS_SAGEMAKER_MP_POST_1_10 = version.parse(SMP_VERSION) >= version.parse("1.10")
else:
    IS_SAGEMAKER_MP_POST_1_10 = False
```
Let's put a name that has Sagemaker inside so a person reading the code later can understand what this is about.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1520,6 +1529,12 @@ def save_pretrained(
         if state_dict is None:
             state_dict = model_to_save.state_dict()
 
+        if is_sagemaker_mp_enabled():
+            # Translate state_dict from smp to hf if saving with smp >= 1.10
+            if SMP_POST_1_10:
+                for smp_to_hf, _ in smp.state.module_manager.translate_functions:
+                    state_dict = smp_to_hf(state_dict)
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1536]:
```suggestion
        if IS_SAGEMAKER_MP_POST_1_10:
            # Translate state_dict from smp to hf if saving with smp >= 1.10
            for smp_to_hf, _ in smp.state.module_manager.translate_functions:
                state_dict = smp_to_hf(state_dict)
```

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -2107,7 +2122,6 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 )
                 raise
         elif from_pt:
-
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2110]: Loos unrelated to SMP, so let's leave it as before (feel free to make another PR focused on style changes!)

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -188,6 +188,10 @@
 
 if is_sagemaker_mp_enabled():
     import smdistributed.modelparallel.torch as smp
+    from smdistributed.modelparallel import __version__ as SMP_VERSION
+
+    SMP_PRE_1_10 = version.parse(SMP_VERSION) < version.parse("1.10")
+    SMP_POST_1_10 = not SMP_PRE_1_10
```
Comment (sgugger) [inline src/transformers/trainer.py:194]:
```suggestion
    IS_SAGEMAKER_MP_POST_1_10 = version.parse(SMP_VERSION) >= version.parse("1.10")
else:
    IS_SAGEMAKER_MP_POST_1_10 = False
```
We don't need two global constants since we can easily use `not` in a test.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -499,14 +503,23 @@ def __init__(
...
-            # When there's mismatch between SMP config and trainer argument, use SMP config as truth
-            if args.fp16 != smp.state.cfg.fp16:
-                logger.warning(
-                    f"FP16 provided in SM_HP_MP_PARAMETERS is {smp.state.cfg.fp16},"
-                    f"but FP16 provided in trainer argument is {args.fp16},"
-                    f"setting to {smp.state.cfg.fp16}"
-                )
-                args.fp16 = smp.state.cfg.fp16
+
+            if SMP_POST_1_10:
```
Comment (sgugger) [inline src/transformers/trainer.py:507]:
```suggestion
            if IS_SAGEMAKER_MP_POST_1_10:
```

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -952,10 +965,23 @@ def create_optimizer_and_scheduler(self, num_training_steps: int):
...
+            else:
+                self.create_scheduler(
+                    num_training_steps=num_training_steps,
+                    optimizer=self.optimizer,
+                )
+        else:
+            self.create_scheduler(
+                num_training_steps=num_training_steps,
+                optimizer=self.optimizer,
+            )
```
Comment (sgugger) [inline src/transformers/trainer.py:984]: This is way too complicated for just a change in the optimizer used:
```suggestion
        if IS_SAGEMAKER_MP_POST_1_10 and smp.state.cfg.fp16:
            optimizer = self.optimizer.optimizer
        else:
            optimizer = self.optimizer
        self.create_scheduler(num_training_steps=num_training_steps, optimizer=optimizer)
```

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -1002,7 +1028,6 @@ def create_optimizer(self):
 
         if is_sagemaker_mp_enabled():
             self.optimizer = smp.DistributedOptimizer(self.optimizer)
-
```
Comment (sgugger) [inline src/transformers/trainer.py:1005]: Same as before, let's remove unrelated changes for those empty lines.

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -1817,14 +1840,34 @@ def _load_from_checkpoint(self, resume_from_checkpoint, model=None):
             # will be resumed in deepspeed_init
             pass
         elif os.path.isfile(os.path.join(resume_from_checkpoint, WEIGHTS_NAME)):
-            # We load the model state dict on the CPU to avoid an OOM error.
-            state_dict = torch.load(os.path.join(resume_from_checkpoint, WEIGHTS_NAME), map_location="cpu")
             # If the model is on the GPU, it still works!
-            load_result = model.load_state_dict(state_dict, strict=strict_load)
+            if strict_load:
```
Comment (sgugger) [inline src/transformers/trainer.py:1844]: Had to dig into the code to realise `strict_load` is just `is_sagemaker_mp_enabled()`. So let's remove the definition at line 1797/1820 and just use
```suggestion
            if is_sagemaker_mp_enabled():
```
here so the code is easier to read.

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -1817,14 +1840,34 @@ def _load_from_checkpoint(self, resume_from_checkpoint, model=None):
...
+                    # Required for smp to not auto-translate state_dict from hf to smp (is already smp).
+                    state_dict["_smp_is_partial"] = False
+                    load_result = model.load_state_dict(state_dict, strict=strict_load)
+                    del state_dict
+            else:
+                # We load the model state dict on the CPU to avoid an OOM error.
+                state_dict = torch.load(os.path.join(resume_from_checkpoint, WEIGHTS_NAME), map_location="cpu")
+                load_result = model.load_state_dict(state_dict, strict=strict_load)
+                del state_dict
             if not strict_load:
```
Comment (sgugger) [inline src/transformers/trainer.py:1868]:
```suggestion

```

File: src/transformers/trainer.py [hunk 10]
```diff
@@ -1817,14 +1840,34 @@ def _load_from_checkpoint(self, resume_from_checkpoint, model=None):
...
+                        )
+                    state_dict = torch.load(os.path.join(resume_from_checkpoint, WEIGHTS_NAME), map_location="cpu")
+                    # Required for smp to not auto-translate state_dict from hf to smp (is already smp).
+                    state_dict["_smp_is_partial"] = False
+                    load_result = model.load_state_dict(state_dict, strict=strict_load)
+                    del state_dict
+            else:
+                # We load the model state dict on the CPU to avoid an OOM error.
+                state_dict = torch.load(os.path.join(resume_from_checkpoint, WEIGHTS_NAME), map_location="cpu")
+                load_result = model.load_state_dict(state_dict, strict=strict_load)
```
Comment (sgugger) [inline src/transformers/trainer.py:1866]:
```suggestion
                load_result = model.load_state_dict(state_dict)
```

File: src/transformers/trainer.py [hunk 11]
```diff
@@ -1817,14 +1840,34 @@ def _load_from_checkpoint(self, resume_from_checkpoint, model=None):
...
+                    # If the 'user_content.pt' file does NOT exist, load with the old smp api.
+                    # Checkpoint must have been saved with the old smp api.
+                    if hasattr(self.args, "fp16") and smp.state.cfg.fp16 is True:
+                        logger.warning(
+                            "Enabling FP16 and loading from smp < 1.10 checkpoint together is not suppported."
+                        )
+                    state_dict = torch.load(os.path.join(resume_from_checkpoint, WEIGHTS_NAME), map_location="cpu")
+                    # Required for smp to not auto-translate state_dict from hf to smp (is already smp).
+                    state_dict["_smp_is_partial"] = False
+                    load_result = model.load_state_dict(state_dict, strict=strict_load)
```
Comment (sgugger) [inline src/transformers/trainer.py:1861]:
```suggestion
                    load_result = model.load_state_dict(state_dict, strict=True)
```

File: src/transformers/trainer.py [hunk 12]
```diff
@@ -1856,10 +1898,28 @@ def _load_best_model(self):
...
+                            tag=WEIGHTS_NAME,
+                            partial=False,
+                            load_optimizer=False,
+                        )
+                    else:
+                        # If the 'user_content.pt' file does NOT exist, load with the old smp api.
+                        # Checkpoint must have been saved with the old smp api.
+                        state_dict = torch.load(best_model_path, map_location="cpu")
+                        state_dict["_smp_is_partial"] = False
+                        load_result = model.load_state_dict(state_dict, strict=strict_load)
```
Comment (sgugger) [inline src/transformers/trainer.py:1916]:
```suggestion
                        load_result = model.load_state_dict(state_dict, strict=True)
```

File: src/transformers/trainer.py [hunk 13]
```diff
@@ -1856,10 +1898,28 @@ def _load_best_model(self):
...
+                else:
+                    # We load the model state dict on the CPU to avoid an OOM error.
+                    state_dict = torch.load(best_model_path, map_location="cpu")
+                    # If the model is on the GPU, it still works!
+                    load_result = model.load_state_dict(state_dict, strict=strict_load)
```
Comment (sgugger) [inline src/transformers/trainer.py:1921]:
```suggestion
                    load_result = model.load_state_dict(state_dict)
```

File: src/transformers/trainer.py [hunk 14]
```diff
@@ -1856,10 +1898,28 @@ def _load_best_model(self):
...
+
                 if not strict_load:
```
Comment (sgugger) [inline src/transformers/trainer.py:1923]:
```suggestion

```

File: src/transformers/trainer.py [hunk 15]
```diff
@@ -1873,7 +1933,6 @@ def _load_best_model(self):
             )
 
     def _issue_warnings_after_load(self, load_result):
-
```
Comment (sgugger) [inline src/transformers/trainer.py:1876]: Another unrelated change.

File: src/transformers/trainer.py [hunk 16]
```diff
@@ -2004,14 +2063,14 @@ def _save_checkpoint(self, model, trial, metrics=None):
                 reissue_pt_warnings(caught_warnings)
         elif is_sagemaker_mp_enabled():
             opt_state_dict = self.optimizer.local_state_dict(gather_if_shard=False)
-            smp.barrier()
```
Comment (sgugger) [inline src/transformers/trainer.py:2007]: Won't this break the older versions?

File: src/transformers/trainer.py [hunk 17]
```diff
@@ -2110,13 +2169,23 @@ def _load_optimizer_and_scheduler(self, checkpoint):
...
-                        )
+                    else:
+                        # Optimizer checkpoint was saved with smp < 1.10
+                        def opt_load_hook(mod, opt):
+                            if SMP_PRE_1_10:
+                                opt.load_state_dict(smp.load(os.path.join(checkpoint, OPTIMIZER_NAME), partial=True))
+                            elif SMP_POST_1_10:
+                                opt.load_state_dict(
+                                    smp.load(os.path.join(checkpoint, OPTIMIZER_NAME), partial=True, back_compat=True)
+                                )
```
Comment (sgugger) [inline src/transformers/trainer.py:2185]:
```suggestion
                            if IS_SAGEMAKER_MP_POST_1_10:
                                opt.load_state_dict(
                                    smp.load(os.path.join(checkpoint, OPTIMIZER_NAME), partial=True, back_compat=True)
                                )
                            else:
                                opt.load_state_dict(smp.load(os.path.join(checkpoint, OPTIMIZER_NAME), partial=True))
```

File: src/transformers/trainer.py [hunk 18]
```diff
@@ -2413,17 +2482,21 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
 
         Will only save from the main process.
         """
-
```
Comment (sgugger) [inline src/transformers/trainer.py:2416]: Another line to leave as was before.

File: src/transformers/trainer.py [hunk 19]
```diff
@@ -2413,17 +2482,21 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
...
         elif is_sagemaker_mp_enabled():
             # Calling the state_dict needs to be done on the wrapped model and on all processes.
+            os.makedirs(output_dir, exist_ok=True)
             state_dict = self.model_wrapped.state_dict()
             if self.args.should_save:
                 self._save(output_dir, state_dict=state_dict)
+            if SMP_POST_1_10:
+                # 'user_content.pt' indicates model state_dict saved with smp >= 1.10
+                user_content = {"smp_ver : >= 1.10"}
+                torch.save(user_content, os.path.join(output_dir, "user_content.pt"))
```
Comment (sgugger) [inline src/transformers/trainer.py:2499]: Why save a dictionary with `torch.save`? There are no tensors in this. Also you can just touch the file as it's not read in the tests before.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. I left some comments on the code.

Please revert all changes unrelated to your PR (deletions of empty lines). They can go in another PR dedicated to styling if you feel strongly about them.

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (viclzhu) [reply src/transformers/modeling_utils.py:97]: Sounds good to me!

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (viclzhu) [reply src/transformers/modeling_utils.py:2110]: Yeah sounds good. These line removals/additions were essentially done on accident while I was going through the code!

File: src/transformers/trainer.py (hunk 4 above)
Comment (viclzhu) [reply src/transformers/trainer.py:194]: Sounds good to me!

File: src/transformers/trainer.py (hunk 6 above)
Comment (viclzhu) [reply src/transformers/trainer.py:984]: Oh thank you! This is much nicer thanks!

File: src/transformers/trainer.py (hunk 19 above)
Comment (viclzhu) [reply src/transformers/trainer.py:2499]: Yeah, this is true!

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (viclzhu) [reply src/transformers/modeling_utils.py:2110]: Oh actually, I think many of these line changes were done by running `black --preview` on the file (as indicated in the contributor guidelines). Should I just not run it then? Thanks!

Comment (sgugger) [reply src/transformers/modeling_utils.py:2110]: If you make sure you have the version of black pinned by Transformers (22.3), you shouldn't have those.

Comment (viclzhu) [reply src/transformers/modeling_utils.py:2110]: Oh yeah my bad, I was using black 22.6!

**Comment by sgugger** [comment]
It looks like GitHub did not like this rebase as it know shows 290 files changed. Could you open a clean new PR after fixing the merge conflicts?

**Comment by viclzhu** [comment]
New clean PR at #18221!
